src/hotspot/cpu/riscv/stubGenerator_riscv.cpp

```diff
@@ -3661,6 +3661,915 @@ class StubGenerator: public StubCodeGenerator {
+    //
+    // VECTORS
+    //  v10 - v13 (512-bits / 4*128 bits / 4*4*32 bits), hold the message
+    //             schedule words (Wt). They start with the message block
+    //             content (W0 to W15), then further words in the message
+    //             schedule generated via vsha2ms from previous Wt.
+    //   Initially:
+    //     v10 = W[  3:0] = { W3,  W2,  W1,  W0}
+    //     v11 = W[  7:4] = { W7,  W6,  W5,  W4}
+    //     v12 = W[ 11:8] = {W11, W10,  W9,  W8}
+    //     v13 = W[15:12] = {W15, W14, W13, W12}
+    //
+    //  v16 - v17 hold the working state variables (a, b, ..., h)
+    //    v16 = {a[t],b[t],e[t],f[t]}
+    //    v17 = {c[t],d[t],g[t],h[t]}
+    //   Initially:
+    //    v16 = {H5i-1, H4i-1, H1i-1 , H0i-1}
+    //    v17 = {H7i-i, H6i-1, H3i-1 , H2i-1}
+    //
+    //  v0 = masks for vrgather/vmerge. Single value during the 16 rounds.
+    //
+    //  v14 = temporary, Wt+Kt
+    //  v15 = temporary, Kt
+    //
+    //  v18/v19 = temporaries, in the epilogue, to re-arrange
+    //            and byte-swap v16/v17
+    //
+    //  v26/v27 = hold the initial values of the hash, byte-swapped.
+    //
+    //  v30/v31 = used to generate masks, vrgather indices.
+    //
+    // During most of the function the vector state is configured so that each
+    // vector is interpreted as containing four 32 bits (e32) elements (128 bits).
+
+    // Set vectors as 4 * 32 bits
+    //
+    // e32: vector of 32b/4B elements
+    // m1: LMUL=1
+    // ta: tail agnostic (don't care about those lanes)
+    // ma: mask agnostic (don't care about those lanes)
+    // x0 is not written, we known the number of vector elements, 8.
+    __ vsetivli(x0, 4, Assembler::e32, Assembler::m1, Assembler::ma, Assembler::ta);
+
+    // Load H[0..8] to produce
+    //  v16 = {a,b,e,f}
+    //  v17 = {c,d,g,h}
+    __ vle32_v(v16, state);                          // v16 = {d,c,b,a}
+    __ addi(state, state, 16);
+    __ vle32_v(v17, state);                          // v17 = {h,g,f,e}
+
+    __ vid_v(v30);                                   // v30 = {3,2,1,0}
+    __ vxor_vi(v30, v30, 0x3);                       // v30 = {0,1,2,3}
+    __ vrgather_vv(v26, v16, v30);                   // v26 = {a,b,c,d}
+    __ vrgather_vv(v27, v17, v30);                   // v27 = {e,f,g,h}
+    __ vmsgeu_vi(v0, v30, 2);                        // v0  = {f,f,t,t}
+    // Copy elements [3..2] of v26 ({d,c}) into elements [3..2] of v17.
+    __ vslideup_vi(v17, v26, 2);                     // v17 = {c,d,_,_}
+    // Merge elements [1..0] of v27 ({g,h}) into elements [1..0] of v17
+    __ vmerge_vvm(v17, v17, v27);                    // v17 = {c,d,g,h}
+    // Copy elements [1..0] of v27 ({f,e}) into elements [1..0] of v16.
+    __ vslidedown_vi(v16, v27, 2);                   // v16 = {_,_,e,f}
+    // Merge elements [3..2] of v26 ({a,b}) into elements [3..2] of v16
+    __ vmerge_vvm(v16, v26, v16);                    // v16 = {a,b,e,f}
```

Comment:
I see the openssl version makes use of index-load to get {f,e,b,a},{h,g,d,c} pre-loop and index-store to put {f,e,b,a},{h,g,d,c} back to {a,b,c,d},{e,f,g,h} post-loop, which is much simpler than this implementation. Please consider.

[1] https://github.com/openssl/openssl/blob/master/crypto/sha/asm/sha256-riscv64-zvkb-zvknha_or_zvknhb.pl#L124-L142

Reply:
The vsetivli is often expensive:ish,  the code in openssl sets it five times before reaching first round.
That don't seem like a good idea, now vsetivli make the code much easier to read yes...

I guess I need to check numbers for that also.. :)

Reply:
Yeah. Why not consider something simpler if there is no known big difference on performance numbers? And this is the first version when RVV-1.0 compatible hardwares are not popular yet :-)

Reply:
Fixed